// llamaindex/dist/engines/query/RetrieverQueryEngine.js
import { BaseQueryEngine } from "@llamaindex/core/query-engine";
import { getResponseSynthesizer } from "@llamaindex/core/response-synthesizers";
import { extractText } from "@llamaindex/core/utils";
/**
 * A query engine that uses a retriever to query an index and then synthesizes the response.
 */ export class RetrieverQueryEngine extends BaseQueryEngine {
    retriever;
    responseSynthesizer;
    nodePostprocessors;
    constructor(retriever, responseSynthesizer, nodePostprocessors){
        super(async (strOrQueryBundle, stream)=>{
            const nodesWithScore = await this.retrieve(typeof strOrQueryBundle === "string" ? strOrQueryBundle : extractText(strOrQueryBundle));
            if (stream) {
                return this.responseSynthesizer.synthesize({
                    query: typeof strOrQueryBundle === "string" ? {
                        query: strOrQueryBundle
                    } : strOrQueryBundle,
                    nodes: nodesWithScore
                }, true);
            }
            return this.responseSynthesizer.synthesize({
                query: typeof strOrQueryBundle === "string" ? {
                    query: strOrQueryBundle
                } : strOrQueryBundle,
                nodes: nodesWithScore
            });
        });
        this.retriever = retriever;
        this.responseSynthesizer = responseSynthesizer || getResponseSynthesizer("compact");
        this.nodePostprocessors = nodePostprocessors || [];
    }
    _getPrompts() {
        return {};
    }
    _updatePrompts() {}
    _getPromptModules() {
        return {
            responseSynthesizer: this.responseSynthesizer
        };
    }
    async applyNodePostprocessors(nodes, query) {
        let nodesWithScore = nodes;
        for (const postprocessor of this.nodePostprocessors){
            nodesWithScore = await postprocessor.postprocessNodes(nodesWithScore, query);
        }
        return nodesWithScore;
    }
    async retrieve(query) {
        const nodes = await this.retriever.retrieve(query);
        const messageContent = typeof query === "string" ? query : query.query;
        return await this.applyNodePostprocessors(nodes, messageContent);
    }
}
